src/main/java/org/benetech/servicenet/service/impl/OrganizationServiceImpl.java

```diff
@@ -547,6 +549,65 @@ public void reactivate(UUID id) {
         .map(organizationMapper::toDto);
     }
 
+    @Override
+    public Organization cloneOrganizationForServiceProvider(UUID orgId, UserProfile user) {
+        Organization organization = organizationRepository.findOneWithAllEagerAssociationsByIdOrExternalDbId(orgId, null);
+        Set<Service> services = organization.getServices();
+        Set<Location> locations = organization.getLocations();
+        Set<DailyUpdate> dailyUpdates = organization.getDailyUpdates();
+
+        Organization orgClone = new Organization(organization);
+        orgClone.setUserProfiles(Collections.singleton(user));
+        orgClone.setAccount(user.getSystemAccount());
+        user.getOrganizations().add(orgClone);
+        organizationRepository.save(orgClone);
+
+        Set<Service> clonedServices = cloneServices(services, orgClone);
+
+        Set<Location> clonedLocations = cloneLocations(locations, orgClone);
+
+        HashSet<DailyUpdate> clonedDailyUpdate = new HashSet<>();
+        dailyUpdates.forEach((DailyUpdate dailyUpdate) -> {
+            DailyUpdate duClone = new DailyUpdate()
+                .update(dailyUpdate.getUpdate())
+                .createdAt(dailyUpdate.getCreatedAt())
+                .expiry(dailyUpdate.getExpiry())
+                .organization(orgClone);
+            clonedDailyUpdate.add(dailyUpdateService.save(duClone));
+        });
+
+        orgClone.setServices(clonedServices);
+        orgClone.setLocations(clonedLocations);
+        orgClone.setDailyUpdates(clonedDailyUpdate);
+
+        organization.setReplacedBy(orgClone);
+        orgClone.setReplacedBy(organization);
+        return orgClone;
+    }
+
+    @Override
+    public void claimRecords(List<UUID> recordsToClaim) {
+        UserProfile currentUserProfile = userService.getCurrentUserProfile();
+
+        recordsToClaim.forEach((UUID recordId) -> cloneOrganizationForServiceProvider(recordId, currentUserProfile));
+
+        currentUserProfile.setClaimedRecords(true);
+        userService.saveProfile(currentUserProfile);
+    }
+
+    @Override
+    public void unclaimRecord(UUID recordToUnclaim) {
+        Organization orgToUnclaim = organizationRepository.getOne(recordToUnclaim);
+        Organization originalOrg = organizationRepository.getOne(orgToUnclaim.getReplacedBy().getId());
+        originalOrg.setReplacedBy(null);
+
+        UserProfile userProfile = userService.getCurrentUserProfile();
```

Comment:
when the user has no records claimed, shouldn't we set the boolean in userprofile to false? maybe it would be better to have a relation with these organizations and have a boolean (claimedRecords.size() == 0) in dto

Reply:
I think for now there is no need for it. The existing `hasClaimedRecords` boolean indicates if the user has ever done the claiming (this is mainly for the frontend), no other checks needed at this point.